public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts

```diff
@@ -592,40 +595,142 @@ function validateRowsLayout(layout: unknown) {
+  return 'state' in element && 'name' in element.state ? element.state.name : '';
+}
+
+function isVizPanel(element: VizPanel | SceneVariables): element is VizPanel {
+  // FIXME: is there another way to do this?
+  return 'pluginId' in element.state;
+}
+
+function isSceneVariables(element: VizPanel | SceneVariables): element is SceneVariables {
+  // Check for properties unique to SceneVariables but not in VizPanel
+  return !('pluginId' in element.state) && ('variables' in element.state || 'getValue' in element);
+}
 
-  const panelQueries = elementMapReferences.panels.get(elementKey);
-  return panelQueries;
+function isSceneDataQuery(query: SceneDataQuery | QueryVariable): query is SceneDataQuery {
+  return 'refId' in query && !('state' in query);
 }
 
 /**
- * Get the persisted datasource for a query
- * When a query is created it could not have a datasource set
+ * Get the persisted datasource for a query or variable
+ * When a query or variable is created it could not have a datasource set
  * we want to respect that and not overwrite it with the auto assigned datasources
  * resolved in runtime
- * @param query
- * @param queryRunner
- * @param autoAssignedPanelDsRef
- * @returns
+ *
  */
-export function getPersistedDSForQuery(
-  query: SceneDataQuery,
-  queryRunner: SceneQueryRunner,
-  autoAssignedPanelDsRef: Set<string> | undefined
-) {
-  // if the query has a refId and it is in the panelDsReferences then it did NOT have a datasource
-  const hasMatchingRefId = autoAssignedPanelDsRef?.has(query.refId);
-  if (hasMatchingRefId) {
+export function getElementDatasource(
+  element: VizPanel | SceneVariables,
+  queryElement: SceneDataQuery | QueryVariable,
+  type: 'panel' | 'variable',
+  queryRunner?: SceneQueryRunner
+): DataSourceRef | undefined {
+  try {
+    if (type === 'panel') {
+      if (!queryRunner || !isVizPanel(element) || !isSceneDataQuery(queryElement)) {
+        return undefined;
+      }
+      // Get datasource for panel query
+      const autoAssignedRefs = getAutoAssignedDSRef(element, 'panels');
+      return getPersistedDSFor(queryElement, autoAssignedRefs, 'query', queryRunner);
+    }
+
+    if (type === 'variable') {
+      if (!isSceneVariables(element) || isSceneDataQuery(queryElement)) {
+        return undefined;
+      }
+      // Get datasource for variable
+      if (!sceneUtils.isQueryVariable(queryElement)) {
+        return undefined;
+      }
+      const autoAssignedRefs = getAutoAssignedDSRef(element, 'variables');
+      // Important: Only return the datasource if it's not in auto-assigned refs
+      // and if the result would not be an empty object
+      const result = getPersistedDSFor(queryElement, autoAssignedRefs, 'variable');
+      return Object.keys(result || {}).length > 0 ? result : undefined;
+    }
+
     return undefined;
-  }
+  } catch (error) {
+    // If no dashboard scene is found, fall back to the original behavior
```

Comment:
I would not just catch the error here, that is going to be very confusing behavior here. and easily lead to bugs later. 

I tried to fix it by passing a long a reference to the serializer with the ds map to all the stateless serializer functions

Reply:
Indeed, that's a better approach [`a50c8d8` (#102571)](https://github.com/grafana/grafana/pull/102571/commits/a50c8d8000b85299341ffea8b85e5e3eb710377a)